torch/distributed/_composable/fsdp/_fsdp_param_group.py

```diff
@@ -166,13 +234,46 @@ def _to_unsharded(self):
                 fsdp_param.to_unsharded()
             self._sharded_state = ShardedState.UNSHARDED
 
+    # Hook Registration #
+    def _register_post_backward_hook(
+        self, args: Tuple[Any, ...], kwargs: Dict[str, Any]
+    ) -> Tuple[Tuple[Any, ...], Dict[str, Any]]:
+        if not torch.is_grad_enabled():
+            return args, kwargs
+        args_list, args_spec = tree_flatten(args)
+        kwargs_list, kwargs_spec = tree_flatten(kwargs)
```

Comment:
so it seems like every pre-forward we'll need to run the module inputs post backward grad autograd function and do tree flatten/unflatten, do we worry about CPU overhead here or because it's mostly module level the overhead is tolerable?

Reply:
We do worry about CPU overhead, but at least the overhead is proportional to the number of args/kwargs, which is typically not that high. I am not sure there is a way to avoid this overhead because the tensors here are created newly each iteration.

The main place we got burned on CPU overhead was when there was `O(NP)` overhead for world size `N` and `P` `nn.Parameter`s. I think this overhead here should be far less, but we can see as we run more real jobs.